[`ruff`] Check for non-context-manager use of `pytest.raises`, `pytest.warns`, and `pytest.deprecated_call` (`RUF061`)

twentyone212121 opened the pull request:
<!--
Thank you for contributing to Ruff! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title?
- Does this pull request include references to any relevant issues?
-->

This PR aims to close #16605.

## Summary

This PR introduces a new rule (`RUF061`) that detects non-contextmanager usage of `pytest.raises`, `pytest.warns`, and `pytest.deprecated_call`. This pattern is discouraged and [was proposed in flake8-pytest-style](https://github.com/m-burst/flake8-pytest-style/pull/332), but the corresponding PR has been open for over a month without activity.

Additionally, this PR provides an unsafe fix for simple cases where the non-contextmanager form can be transformed into the context manager form. Examples of supported patterns are listed in `RUF061_raises.py`, `RUF061_warns.py`, and `RUF061_deprecated_call.py` test files.

The more complex case from the original issue (involving two separate statements):
```python
excinfo = pytest.raises(ValueError, int, "hello")
assert excinfo.match("^invalid literal")
```
is getting fixed like this:
```python
with pytest.raises(ValueError) as excinfo:
    int("hello")
assert excinfo.match("^invalid literal")
```
Putting match in the raises call requires multi-statement transformation, which I am not sure how to implement.

## Test Plan

<!-- How was it tested? -->

New test files were added to cover various usages of the non-contextmanager form of pytest.raises, warns, and deprecated_call.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -156,6 +159,48 @@ impl Violation for PytestRaisesWithoutException {
+/// ## References
+/// - [`pytest` documentation: `pytest.raises`](https://docs.pytest.org/en/latest/reference/reference.html#pytest-raises)
+#[derive(ViolationMetadata)]
+pub(crate) struct DeprecatedPytestRaisesCallableForm;
```
We might want to leave the `Deprecated` part out of this name since it's not actually deprecated yet, and my quick skim of the open PR making it deprecated made it sound a bit contentious to deprecate it at all. I think this can still be a useful stylistic rule even if it's not formally deprecated.

As more of a nit, I find the rest of the name a bit awkward, but I can't think of anything better that still fits within our naming conventions.

ntBre replied:
To be fair, it is called the `legacy` form in the [docs](https://docs.pytest.org/en/stable/how-to/assert.html#alternate-form-legacy), but they also include this statement:

> Nonetheless, this form is fully supported and not deprecated in any way.

webknjaz replied:
Yeah, we've been debating the deprecation and mostly settled on “let's try soft-deprecating and see if people complain, but meanwhile let the formatters/linters catch this”: https://github.com/pytest-dev/pytest/pull/13241#issuecomment-2694138535. The PR currently uses a `PendingDeprecationWarning` FWIW.

That “legacy” form is the stdlib unittest baggage, I think. It's probably mostly met in ancient code bases. I'd call it “discouraged” instead of “deprecated”. Or, perhaps, “unintuitive” or “callback-style”.

twentyone212121 replied:
Should I rename it to `LegacyPytestRaises` or maybe `CallablePytestRaises`?

webknjaz replied:
Callable seems off. Callback would be more accurate.

MichaReiser replied:
How about `LegacyFormPytestRaises` or `PytestRaisesAsFunction`? 

It would match the terminolgoy used in the documentation. Other options are `pytest-raises-with-func` because that's really what we detect. What I find interesting is that this form is listed in context manager form in the pytest documentation (but what I understand is that it isn't supposed to be used in a context manager position)

with raises(expected_exception: [type](https://docs.python.org/3/library/functions.html#type)[E] | [tuple](https://docs.python.org/3/library/stdtypes.html#tuple)[[type](https://docs.python.org/3/library/functions.html#type)[E], ...], func: [Callable](https://docs.python.org/3/library/typing.html#typing.Callable)[[...], [Any](https://docs.python.org/3/library/typing.html#typing.Any)], *args: [Any](https://docs.python.org/3/library/typing.html#typing.Any), **kwargs: [Any](https://docs.python.org/3/library/typing.html#typing.Any)) → [ExceptionInfo](https://docs.pytest.org/en/stable/reference/reference.html#pytest.ExceptionInfo)[E] as excinfo

ntBre replied:
I like `LegacyFormPytestRaises` or maybe `PytestRaisesLegacyForm`, but these all sound good to me.

Ahh, that does explain why I didn't see a signature under the `Legacy` heading. I think you're right that that is the legacy form signature but with `with` in front of it.

What do you think about the rule category? `RUF` might make more sense just to be safe. The upstream PR still hasn't had any updates, but new rules were added in January, so it's still active.

MichaReiser replied:
We should add the rule to the Ruff group.

twentyone212121 replied:
I also like `LegacyFormPytestRaises`, so if everyone agrees, I'll rename the rule

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -156,6 +159,48 @@ impl Violation for PytestRaisesWithoutException {
+pub(crate) struct DeprecatedPytestRaisesCallableForm;
+
+impl Violation for DeprecatedPytestRaisesCallableForm {
+    const FIX_AVAILABILITY: ruff_diagnostics::FixAvailability =
```
nit: I'd just import this so it fits on one line.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -272,3 +351,128 @@ fn exception_needs_match(checker: &Checker, exception: &Expr) {
     }
 }
+
+fn try_fix_legacy_raises(stmt: &Stmt, semantic: &SemanticModel) -> anyhow::Result<StmtWith> {
```
Does anything here actually return a `Result` naturally? I might be missing something, but everything looks like it returns an `Option`, so we could just return an `Option` from the function and still use `?` everywhere. I don't think the additional messages from `bail` and `context` justify using a `Result` here.

twentyone212121 replied:
I agree, I replaced `Result` with `Option`. The additional messages were helpful for debugging earlier, but using `Option` makes things clearer now.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -272,3 +351,128 @@ fn exception_needs_match(checker: &Checker, exception: &Expr) {
+    legacy_raises_call: &ast::ExprCall,
+    match_arg: Option<&Expr>,
+    optional_vars: Option<&Expr>,
+) -> anyhow::Result<StmtWith> {
```
Same comment as above, I'd return an `Option` here if possible.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -272,3 +351,128 @@ fn exception_needs_match(checker: &Checker, exception: &Expr) {
+    match_arg: Option<&Expr>,
+    optional_vars: Option<&Expr>,
+) -> anyhow::Result<StmtWith> {
+    let expected_exception = match legacy_raises_call
+        .arguments
+        .find_argument("expected_exception", 0)
```
If you want the value, I think you can use `find_argument_value` instead of having to `match` manually.

twentyone212121 replied:
Thanks! That’s definitely cleaner

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -272,3 +351,128 @@ fn exception_needs_match(checker: &Checker, exception: &Expr) {
+
+    let func = match legacy_raises_call
+        .arguments
+        .find_argument("func", 1)
```
Same as above, `find_argument_value`.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -272,3 +351,128 @@ fn exception_needs_match(checker: &Checker, exception: &Expr) {
+        },
+    };
+
+    let func_args = legacy_raises_call
+        .arguments
+        .args
+        .iter()
+        .filter(|&arg| arg != expected_exception && arg != func)
+        .map(Expr::clone)
+        .collect();
+
+    let func_keywords = legacy_raises_call
+        .arguments
+        .keywords
+        .iter()
+        .filter(|&keyword| &keyword.value != expected_exception && &keyword.value != func)
+        .map(ast::Keyword::clone)
+        .collect();
```
I'm not sure there's anything inherently wrong with this, but it feels weird to be comparing the actual `Expr` values for equality here. Can we just use `arguments_source_order` and skip the first two elements of the iterator?

https://github.com/astral-sh/ruff/blob/1108f662e47e8ff5456f00e5a0742f1cd9e95852/crates/ruff_python_ast/src/nodes.rs#L2904-L2908

We should have already bailed out if we didn't find the `func` or `expected_exception` arguments, I think.

You'll still have to partition these into `args` and `keywords` to put into `func_call` below, but I think it will read a bit more nicely. You can also use `.cloned()` on the iterator to avoid needing to map separate `clone` calls.

ntBre commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -185,6 +230,40 @@ pub(crate) fn raises_call(checker: &Checker, call: &ast::ExprCall) {
+                let generated = try_fix_legacy_raises(stmt, checker.semantic()).map(|with| {
+                    let generated = checker.generator().stmt(&Stmt::With(with));
+                    let first_line = checker.locator().line_str(stmt.start());
+                    let indentation = leading_indentation(first_line);
```
This may not be helpful because I'm not that familiar with this part of the API, but we do have an `adjust_indentation` function in the `edits` module:

https://github.com/astral-sh/ruff/blob/1108f662e47e8ff5456f00e5a0742f1cd9e95852/crates/ruff_linter/src/fix/edits.rs#L348-L354

It might be worth giving it a try to avoid having to manually handle it.

twentyone212121 replied:
From what I could tell, `adjust_indentation` seems to be mostly used for dedenting code that already exists in the source.

When I was looking into similar fixes that turn a single statement into an indented block, I found the [lambda-assignment rule](https://docs.astral.sh/ruff/rules/lambda-assignment/), and borrowed the indentation logic from there:
https://github.com/astral-sh/ruff/blob/1108f662e47e8ff5456f00e5a0742f1cd9e95852/crates/ruff_linter/src/rules/pycodestyle/rules/lambda_assignment.rs#L80-L97

ntBre commented on crates/ruff_linter/src/codes.rs:
```diff
@@ -839,6 +839,7 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
         (Flake8PytestStyle, "029") => (RuleGroup::Preview, rules::flake8_pytest_style::rules::PytestWarnsWithoutWarning),
         (Flake8PytestStyle, "030") => (RuleGroup::Preview, rules::flake8_pytest_style::rules::PytestWarnsTooBroad),
         (Flake8PytestStyle, "031") => (RuleGroup::Preview, rules::flake8_pytest_style::rules::PytestWarnsWithMultipleStatements),
+        (Flake8PytestStyle, "032") => (RuleGroup::Preview, rules::flake8_pytest_style::rules::DeprecatedPytestRaisesCallableForm),
```
As you said in the PR description, this may end up needing to be a ruff rule instead, but we can delay changing that until right before landing.

twentyone212121 replied:
Should I assign `RUF103` code to the rule? Also, if the rule is renamed, is it okay for the checking code to be in `flake8_pytest_style` folder?

ntBre replied:
It looks like there are two open PRs for rules using RUF060 (and an additional closed one), so let's go for `RUF061`, which I didn't see any results for.

ntBre reviewed (commented):
Thanks for working on this!

I had some code style suggestions, but the overall approach looks good to me.

With regard to the multi-statement case, I think your example in the PR summary _is_ actually ideal, based on my reading of the pytest docs. See the first `Note` box in [this section](https://docs.pytest.org/en/stable/reference/reference.html#pytest-raises): 

> When using pytest.raises as a context manager, it’s worthwhile to note that normal context manager rules apply and that the exception raised must be the final line in the scope of the context manager. Lines of code after that, within the scope of the context manager will not be executed. For example:

However, to be safe, we may want to avoid offering a fix if the value returned by `raises` is assigned to a variable, or at least if we can see that variable being used later. 
On the other hand, it's already an unsafe fix, so if my reading of the pytest docs is correct, the behavior here might be right in most cases.

Either way, we should add some test cases demonstrating whatever the behavior ends up being.

twentyone212121 commented:
Thanks for the review!

About the multi-statement case — I initially thought it wasn’t ideal, since in the context manager form, `pytest.raises` accepts the match argument directly, so there's no need for a separate assertion. That said, I’m not sure whether this rule should be responsible for transforming that pattern.

Regarding the tests, I’ve added a multi-statement case, as well as a case where `func` is a lambda.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pytest_style/rules/raises.rs:
```diff
@@ -185,6 +230,40 @@ pub(crate) fn raises_call(checker: &Checker, call: &ast::ExprCall) {
+                    let indentation = leading_indentation(first_line);
+                    let mut indented = String::new();
+                    for (idx, line) in generated.universal_newlines().enumerate() {
+                        if idx == 0 {
```
Is it guaranteed that the generator always fits the with-context on a single line?

twentyone212121 replied:
From my testing, the generator always produces a single-line `with` statement, even if the original `pytest.raises` call was split across multiple lines. I tried two cases:

1. A very long `expected_exception`:
   ```python
   def test():
       pytest.raises(
           VeryLongExceptionThatExceeds80CharactersToTestWhetherGeneratedWithStatementIsOneLineOrNot,
           test_func,
       )
   ```
   After applying the rule, it’s fixed to:
   ```python
   def test():
       with pytest.raises(VeryLongExceptionThatExceeds80CharactersToTestWhetherGeneratedWithStatementIsOneLineOrNot):
           test_func()
   ```
   Although the resulting line is not formatted according to ruff, it is still generated as a single line.

2. A multiline `match` string:
   ```python
   def test():
       pytest.raises(ValueError, test_func).match(
           """a
           b"""
       )
   ```
   After applying the rule:
   ```python
   def test():
       with pytest.raises(ValueError, match="""a\n           b"""):
           test_func()
   ```
   The `with` statement is produced on a single line.

I don't know if it's okay to generate potentially unformatted code, but it seems like the generator doesn't format its output.

ntBre reviewed (commented):
Thanks, this looks great. I think now we just need to:
- decide on a name
- decide on a rule category
- address Micha's comment about the single line

I made some suggestions about the first two [here](https://github.com/astral-sh/ruff/pull/17368#discussion_r2059015744), but I'm definitely interested in other thoughts.

twentyone212121 commented:
Should we also handle legacy forms of `pytest.warns` and `pytest.deprecated_call`? They are mentioned in the [PR to pytest](https://github.com/pytest-dev/pytest/pull/13241). These functions are slightly different from `pytest.raises`, so I’ll need to think about how best to integrate them into the current code to avoid code duplication.

ntBre commented:
> Should we also handle legacy forms of `pytest.warns` and `pytest.deprecated_call`? They are mentioned in the [PR to pytest](https://github.com/pytest-dev/pytest/pull/13241). These functions are slightly different from `pytest.raises`, so I’ll need to think about how best to integrate them into the current code to avoid code duplication.

I think it's okay to handle these separately in follow-up PRs, or even as different rules, if that sounds good to @MichaReiser.

If we want to include them in this rule, it could possibly affect our name choice too.

ntBre commented:
> Whether it should be one or multiple rules normally comes down to whether there are cases where a user might only want to be warned about `pytest.warns` but not about `pytest.deprecated_call` or `pytest.raises`.

From that perspective, I'd probably lean towards one rule because it seems more consistent stylistically to move them all at once. But I'm definitely open to other ideas. I've only personally used the context manager forms of these.

We can still add support for the other functions in follow-up PRs (unless they're easy to add here), but we should definitely decide on the scope and name here to avoid future churn with the name, as Micha said.

twentyone212121 commented:
I agree with Brent that it makes sense to combine all of these into 1 rule. I've made the following changes to the PR:

- In the first commit, I renamed the rule and moved it to the `RUF` category.
- In the second one, I've expanded the rule to cover `pytest.warns` and `pytest.deprecated_call`. Also, renamed it to `LegacyFormPytestRaisesWarnsDeprecatedCall` (I'm open to suggestions about the name). 

Note that unlike `pytest.raises`, both `pytest.warns` and `pytest.deprecated_call` return the result of the `func` they execute, not a context. Therefore, if these functions' results are assigned to a target, we treat it as an assign target in the body rather than an optional var. I tried to minimize code duplication by handling all variants in the same functions, which may make the code appear somewhat complex.

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+        call.range(),
+    );
+
+    let stmt = checker.semantic().current_statement();
```
```suggestion
    let stmt = semantic.current_statement();
```

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+    if !has_leading_content(stmt.start(), checker.source())
+        && !has_trailing_content(stmt.end(), checker.source())
+    {
+        if let Some(with_stmt) = try_fix_legacy_call(context_type, stmt, checker.semantic()) {
```
```suggestion
        if let Some(with_stmt) = try_fix_legacy_call(context_type, stmt, semantic) {
```

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+        Stmt::Expr(StmtExpr { value, .. }) => {
+            let call = value.as_call_expr()?;
+
+            if PytestContextType::from_expr_name(&call.func, semantic) == Some(context_type) {
+                generate_with_statement(context_type, call, None, None, None)
+            } else if let PytestContextType::Raises = context_type {
```
I think some example Python code in a comment might help me here. It seems like this is doing something important, but all I can think is "didn't we already call `PytestContextType::from_expr_name` above?" What kind of code will lead to the different branches here?

twentyone212121 replied:
This handles two cases for legacy calls:
1. Direct usage: `pytest.raises(ZeroDivisionError, func, 1, b=0)`
2. With match method: `pytest.raises(ZeroDivisionError, func, 1, b=0).match("division by zero")`

The second branch specifically looks for `raises().match()` pattern which only exists for `raises` (not `warns`/`deprecated_call`) since only `raises` returns an `ExceptionInfo` object with a `.match()` method. We need to preserve this match condition when converting.

I have modified the comment to explain it better. There is also a test case for this in `RUF061_raises.py`:
```Python
def test_error_match():
    pytest.raises(ZeroDivisionError, func, 1, b=0).match("division by zero")
```

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+    optional_vars: Option<&Expr>,
+    assign_targets: Option<&[Expr]>,
+) -> Option<StmtWith> {
+    let expected = if let Some((name, position)) = context_type.expected_arg() {
+        Some(legacy_call.arguments.find_argument_value(name, position)?)
+    } else {
+        None
+    };
```
I think we could use `and_then` here (but haven't tested it):

```suggestion
    let expected = context_type.expected_arg().and_then(|(name, position)| legacy_call.arguments.find_argument_value(name, position));
```

twentyone212121 replied:
I think we can't use `and_then` here because we need different behavior:
For `raises`/`warns`, the `?` operator returns early if the argument isn't found.
For `deprecated_call`, we expect `None` and continue.

ntBre commented on crates/ruff_linter/src/rules/ruff/snapshots/ruff_linter__rules__ruff__tests__RUF061_RUF061_deprecated_call.py.snap:
```diff
@@ -0,0 +1,57 @@
+24 24 | def test_error_lambda():
+25    |-    pytest.deprecated_call(lambda: warnings.warn("", DeprecationWarning))
+   25 |+    with pytest.deprecated_call():
+   26 |+        (lambda: warnings.warn("", DeprecationWarning))()
```
Hmm, it would be kind of nice if we could unwrap the `lambda` since this looks a bit awkward, but I'm not sure how tricky that would be.

twentyone212121 replied:
Yes, unwrapping the lambda would be cleaner output. For simple cases like this with no arguments, we could extract the lambda body and use it directly.

However, I am not sure if it's within the scope of the current rule. If you think it's worth addressing now though, I can expand the implementation.

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    fn name(self) -> &'static str {
+        match self {
+            Self::Raises => "raises",
+            Self::Warns => "warns",
+            Self::DeprecatedCall => "deprecated_call",
+        }
+    }
```
Could we do a `Display` impl here? Then we could possibly store a `PytestContextType` on the `Violation` struct and just pass it directly to the `format!` calls above.

ntBre commented on crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs:
```diff
@@ -0,0 +1,294 @@
+/// - [`pytest` documentation: `pytest.warns`](https://docs.pytest.org/en/latest/reference/reference.html#pytest-warns)
+/// - [`pytest` documentation: `pytest.deprecated_call`](https://docs.pytest.org/en/latest/reference/reference.html#pytest-deprecated-call)
+#[derive(ViolationMetadata)]
+pub(crate) struct LegacyFormPytestRaisesWarnsDeprecatedCall {
```
This name seems a bit unwieldy, but I'm not really coming up with something better. Is there some kind of catch-all term that covers all three? 

The pytest docs for `warns` actually say:

> Assert that code **raises** a particular class of warning.

So I think it would be fairly reasonable just to call it `LegacyFormPytestRaises` still. Not to say it's an ideal practice, but we do have some other rules with names that are more specific than the actual implementation, like [yield-outside-function (F704)](https://docs.astral.sh/ruff/rules/yield-outside-function/#yield-outside-function-f704), which actually covers `yield`, `yield from`, and even `await`.

twentyone212121 replied:
Yes, it's quite long. I agree that `LegacyFormPytestRaises` will be cleaner. I renamed it in the latest commit.

ntBre reviewed (commented):
Thanks! This is looking great. Just a few minor nits/suggestions/questions.

jakkdl commented:
Oh I'd completely missed this PR, happy to see progress despite my PR to `flake8-pytest-style` going stale. :rocket: 

I suppose I can see the logic that transforming
```python
with pytest.raises(ValueError) as excinfo:
    ...
assert excinfo.match("bar")
```
into
```python
with pytest.raises(ValueError, match="bar"):
    ...
```
might not be the primary responsibility of this rule, but I do think it's a shame as the latter is *much* cleaner.

ntBre reviewed (approved):
Thanks and sorry for the delay here. I think all of my questions have been resolved. 

Could you merge `main` and update the title/description? GitHub isn't showing any conflicts, but I've made some changes to our diagnostic reporting that I think will actually cause problems if we don't merge first. Basically instead of `Diagnostic::new` and a later `checker.report_diagnostic`, you'll just need to call `checker.report_diagnostic` with whatever you previously passed to `Diagnostic::new`. I'm happy to help with that or take care of it if you prefer.

Other than that I think this is good to go.

ntBre commented:
(Just made the title a bit more verbose for inclusion directly into the release notes!)
